### PR TITLE
Addition of eslint plugin

### LINKED ISSUE #154: TSLint rules to enforce TypeDoc?
I'm having trouble finding if this exists... I know there are TSLint rules to enforce JSDoc comments, which is what I am using now. Wondering if there are any plans to have a similar rule to enforce complete TSDoc annotations. Also not sure if this question belongs here or in the TSLint repo, but thought the contributors here might have a better idea.

### PR DESCRIPTION
**Pull request by bafolts**
This pull request adds an eslint plugin which validates tsdoc comment blocks. This should resolve #154

### CODE REVIEW

File: eslint-plugin/src/index.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+var messageIds = {};
+
+allTsdocMessageIds.forEach(function (messageId) {
+    messageIds[messageId] = messageId;
+});
+
+module.exports = {
+    rules: {
+        "tsdoc-comments": {
+            meta: {
```
Comment (iclanton) [inline eslint-plugin/src/index.js]: We should include at least the `type` and `docs` properties here (https://github.com/eslint/eslint/blob/master/docs/developer-guide/working-with-rules.md#rule-basics).

File: eslint-plugin/src/index.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                        var commentBlocks = sourceCode.getCommentsBefore(node).filter(function (comment) {
+                            return comment.type === "Block";
+                        });
+                        if (commentBlocks.length > 0) {
+                            var commentBlock = commentBlocks[0];
+                            var commentString = "/*" + commentBlock.value + "*/";
+                            var results = tsDocParser.parseString(commentString).log;
+                            if (results._messages.length > 0) {
+                                results._messages.forEach(function (message) {
+                                    context.report({
```
[the remark this answers is not in the document]
Comment (octogonz) [reply eslint-plugin/src/index.js]: I think we do want separate lint errors for each TSDoc error. This way the red squiggly underlines can be as accurate as possible.

File: eslint-plugin/.npmrc [hunk 3]
```diff
@@ -0,0 +1,2 @@
+registry=https://registry.npmjs.org/
```
[the remark this answers is not in the document]
Comment (octogonz) [reply eslint-plugin/.npmrc]: If so we should also delete the copy under `/tsdoc/.npmrc`.

I think you're right though -- this probably predates the ADO publishing pipeline.

File: eslint-plugin/src/index.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                var tsDocParser = new TSDocParser();
+                var sourceCode = context.getSourceCode();
+                return {
+                    FunctionDeclaration: function (node) {
+                        var commentBlocks = sourceCode.getCommentsBefore(node).filter(function (comment) {
+                            return comment.type === "Block";
+                        });
+                        if (commentBlocks.length > 0) {
+                            var commentBlock = commentBlocks[0];
+                            var commentString = "/*" + commentBlock.value + "*/";
```
[the remark this answers is not in the document]
Comment (octogonz) [reply eslint-plugin/src/index.js]: @iclanton I noticed this also, however I believe that the other `*` is actually inside the `commentBlock.value`.  

For example, if you wrote `/******/` then the AST sees this as a block comment containing `****`.  It doesn't interpret `/**` specially.

File: eslint-plugin/src/index.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            meta: {
+                messages: messageIds
+            },
+            create: function(context) {
+                var tsDocParser = new TSDocParser();
+                var sourceCode = context.getSourceCode();
+                return {
+                    FunctionDeclaration: function (node) {
+                        var commentBlocks = sourceCode.getCommentsBefore(node).filter(function (comment) {
+                            return comment.type === "Block";
```
[the remark this answers is not in the document]
Comment (bafolts) [reply eslint-plugin/src/index.js]: I am not sure, the type declaration has two string types listed.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/estree/index.d.ts#L45

**Comment by bafolts** [comment]
I will make the suggested changes here and comment back when I am done. Thanks for the feedback.

**Comment by bafolts** [comment]
I believe I have made all the suggested improvements. Let me know if there are any more.

File: eslint-plugin/package.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "eslint-plugin-tsdoc",
+  "version": "0.1.0",
+  "description": "eslint plugin for tsdoc",
+  "private": false,
+  "license": "MIT",
+  "scripts": {
+    "build": "tsc",
```
Comment (iclanton) [inline eslint-plugin/package.json]: We probably want to run `eslint` on this package as well. Maybe copy parts of [the build script](https://github.com/microsoft/tsdoc/blob/master/tsdoc/build.js) from `tsdoc`.

File: eslint-plugin/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
+
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: Can you include the copyright notice?

File: eslint-plugin/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import * as ESTree from "estree";
+
+const messageIds: {[x: string]: string} = {};
+
+allTsdocMessageIds.forEach((messageId: string) => {
+    messageIds[messageId] = messageId + ": {{ unformattedText }}";
+});
+
+module.exports = {
+    rules: {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: Can you use two-space indentation?

File: eslint-plugin/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import * as eslint from "eslint";
+import * as ESTree from "estree";
+
+const messageIds: {[x: string]: string} = {};
+
+allTsdocMessageIds.forEach((messageId: string) => {
+    messageIds[messageId] = messageId + ": {{ unformattedText }}";
+});
+
+module.exports = {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: You should just be able to say `exports = ...`.

File: eslint-plugin/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
+
+import { TSDocParser } from "@microsoft/tsdoc";
+import { allTsdocMessageIds } from "@microsoft/tsdoc/lib/parser/TSDocMessageId";
+import * as eslint from "eslint";
+import * as ESTree from "estree";
+
+const messageIds: {[x: string]: string} = {};
+
+allTsdocMessageIds.forEach((messageId: string) => {
+    messageIds[messageId] = messageId + ": {{ unformattedText }}";
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: You can use a template string here:

```suggestion
  messageIds[messageId] = `${messageId}: {{ unformattedText }}`;
```

File: eslint-plugin/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                messages: messageIds,
+                type: "problem",
+                docs: {
+                    description: "Validates tsdoc comments",
+                    category: "Typescript",
+                    recommended: false,
+                    url: "https://github.com/microsoft/tsdoc"
+                }
+            },
+            create: function(context: eslint.Rule.RuleContext) {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: Use a lambda function here.

```suggestion
            create: (context: eslint.Rule.RuleContext) => {
```

File: eslint-plugin/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                var tsDocParser = new TSDocParser();
+                var sourceCode = context.getSourceCode();
+                var checkCommentBlocks = function (node: ESTree.Node) {
+                    var commentBlocks = sourceCode.getCommentsBefore(node).filter(function (comment: ESTree.Comment) {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]:
```suggestion
                    var commentBlocks = sourceCode.getCommentsBefore(node).filter((comment: ESTree.Comment) => {
```

File: eslint-plugin/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+                type: "problem",
+                docs: {
+                    description: "Validates tsdoc comments",
+                    category: "Typescript",
+                    recommended: false,
+                    url: "https://github.com/microsoft/tsdoc"
+                }
+            },
+            create: function(context: eslint.Rule.RuleContext) {
+                var tsDocParser = new TSDocParser();
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]: Most of these should be `const`s, and can you include type annotations?

I think these issues will be caught by the linter.

File: eslint-plugin/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+                var sourceCode = context.getSourceCode();
+                var checkCommentBlocks = function (node: ESTree.Node) {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]:
```suggestion
                var checkCommentBlocks = (node: ESTree.Node) => {
```

File: eslint-plugin/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+                    var commentBlocks = sourceCode.getCommentsBefore(node).filter(function (comment: ESTree.Comment) {
+                        return comment.type === "Block";
+                    });
+                    if (commentBlocks.length > 0) {
+                        var commentBlock = commentBlocks[0];
+                        var commentString = "/*" + commentBlock.value + "*/";
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]:
```suggestion
                        var commentString = `/**${commentBlock.value}*/`;
```

File: eslint-plugin/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+                        var results = tsDocParser.parseString(commentString).log;
+                        for (let message of results.messages) {
```
Comment (iclanton) [inline eslint-plugin/src/index.ts]:
```suggestion
                        for (const message of results.messages) {
```

File: eslint-plugin/src/tests/index.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
+
+import { RuleTester } from "eslint";
+const rule = require("../index").rules["tsdoc-comments"];
```
Comment (iclanton) [inline eslint-plugin/src/tests/index.ts]:
```suggestion
import * as rules from '../index';
const rule: eslint.Rule /* or whatever the type is here */ = rules['tsdoc-comments'];
```

File: eslint-plugin/src/tests/index.ts [hunk 18]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (iclanton) [inline eslint-plugin/src/tests/index.ts]: Include the copyright.

File: eslint-plugin/package.json [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+    "test": "node tests/index.js"
```
Comment (iclanton) [inline eslint-plugin/package.json]: This path should now be `lib/tests/index.js`.

Although it'd probably be best to include this in the build script so we run the test as part of the build.

File: eslint-plugin/package.json [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  "dependencies": {
+    "@microsoft/tsdoc": "0.12.14"
+  },
+  "devDependencies": {
+    "@types/eslint": "^6.1.3",
```
Comment (iclanton) [inline eslint-plugin/package.json]: `@types/` packages don't follow semver. Pin this to the specific version.

File: eslint-plugin/src/index.ts (hunk 9 above)
Comment (iclanton) [inline eslint-plugin/src/index.ts]: It would be useful to ensure that the object we export is well-typed.

**Comment by iclanton** [comment]
Thanks so much for putting this together! This rule will be really useful.

Most of the remaining comments are fairly superficial, and they can probably be addressed in a follow-up PR if you like.

**Comment by bafolts** [comment]
I have made all the requested changes. It appears something with the build is failing in the `playground` directory. If I update the `@types/react-dom` version to 16.9.3 locally I can fix the problem, but for the life of me I can't get `rushstack` to build with the version I specify in `playground/package.json`.

**Comment by octogonz** [comment]
It's [this old issue](https://github.com/yarnpkg/yarn/issues/6695).  The TSDoc repo was serving as a guinea pig for the Yarn package manager, but Yarn really doesn't play well with DefinitelyTyped.  There's an active megathread where they're finally trying to sort this out.  @iclanton I wonder if we should just switch to PNPM for the time being?

I've pushed a fix to your branch.

File: eslint-plugin/.npmrc (hunk 3 above)
[the remark this answers is not in the document]
Comment (iclanton) [reply eslint-plugin/.npmrc]: This probably actually predates this repo's migration to Rush. Let's clean these up in a follow-up PR.

**Pull request merged**